### ISSUE TITLE
VEP: disable the G2P plugin when annotating

### ISSUE THREAD

**Issue by lbeltrame** [bcbio/bcbio-nextgen#2818]
See https://github.com/Ensembl/ensembl-vep/issues/375. At this point here fails because all these connections are blocked by my institution's firewall.

**Comment by roryk** [comment]
Thanks, is there a way you can work around the firewall? Often times you can set a http_proxy or something similar that will let you circumvent it.

If you move **hg38/variation/G2P.csv** then bcbio will skip running G2P since it won't be able to find the file. Let me know if either one of those things gets you going.

**Comment by matthdsm** [comment]
An update to the plugin was planned for the release of vep 96, but has been postponed to v97 due to some technical diffulties. They also plan to patch the v96 plugin if it's ready on time.

Cheers
M

**Comment by audyavar** [comment]
Does VEP and its annotation files get installed by default with BCBIO1.1.6 version?

**Comment by matthdsm** [comment]
Hi @audyavar 
The VEP executable gets installed together with bcbio. To get the cache, you'll need to use the `--datatarget vep` flag during installation.
More info here: https://bcbio-nextgen.readthedocs.io/en/latest/contents/installation.html#customizing-data-installation

Cheers
M

**Comment by lbeltrame** [comment]
Unfortunately not having a plugin is not enough. It needs to be physically removed from the command line or it will still try to connect.

@roryk In my institution IT policies cannot be discussed or changed, unfortunately. I'm stuck with them (and all non standard ports plus all UDP outbound are blocked).

**Comment by ameynert** [comment]
I have the G2P.csv file set, but I'm still not getting any results. Error message:

```
WARNING: Failed to instantiate plugin G2P: ERROR: No file specified or could not read from file file:/home/u027/project/software/bcbio-1.1.3/genomes/Hsapiens/hg38/variation/G2P.csv
```

But it's definitely there and readable:
```
[ameynert@ultra test]$ ls -lh /home/u027/project/software/bcbio-1.1.3/genomes/Hsapiens/hg38/variation/G2P.csv 
-rw-r--r--. 1 ameynert u027 725K May 21 15:23 /home/u027/project/software/bcbio-1.1.3/genomes/Hsapiens/hg38/variation/G2P.csv
```

Is there a way of specifying the G2P file outside of the hg38-resources.yaml? One of my frequent bcbio uses is for rare disease trios, and at the moment I run VEP afterwards for whichever panel has been requested. Ideally I'd like to have VEP/G2P done within the bcbio run, which means specifying it in the YAML for each trio or batch of trios.

**Comment by ameynert** [comment]
Additional note -I have not removed the --offline parameter from bcbio's vep call. I understand G2P won't run at all in this scenario, so perhaps it's just giving me a misleading error message about the file.

**Comment by matthdsm** [comment]
Hi Alison,
Do you have the possibility to upgrade your bcbio install to the latest version? I think that might help with your issues.
`bcbio_nextgen.py upgrade --tools --data`

The error you're experiencing has nothing to do with the `--offline` flag, but points to a filesystem issue. Are you sure the file is available on the systems you run you analysis on?

M

**Comment by ameynert** [comment]
I have upgraded to bcbio 1.1.5 and am encountering the same issue. The file is definitely available.

```
[ameynert@ultra test]$ head stdout_warnings.txt 
WARNING: Failed to instantiate plugin G2P: ERROR: No file specified or could not read from file file:/home/u027/project/software/bcbio/genomes/Hsapiens/hg38/variation/DDG2P.30082018.csv

```

```
[ameynert@ultra test]$ ls -lh /home/u027/project/software/bcbio/genomes/Hsapiens/hg38/variation/DDG2P.30082018.csv
-rw-r--r--. 1 ameynert u027 725K Jun  3 14:32 /home/u027/project/software/bcbio/genomes/Hsapiens/hg38/variation/DDG2P.30082018.csv
```

**Comment by matthdsm** [comment]
Hi Allison,

I notice you've put the file there manually. The filename for G2P is managed bij the bcbio installer and cloudbiolinux. If you rename `DDG2P.30082018.csv` to `G2P.csv`, it should work.

https://github.com/chapmanb/cloudbiolinux/blob/master/ggd-recipes/hg38/genotype2phenotype.yaml#L23

M

**Comment by ameynert** [comment]
I changed the filename in genomes/Hsapiens/hg38/seq/hg38-resources.yaml as well. VEP was using the correct file:

```
[2019-06-04T10:19Z] unset PERL5LIB && export PATH=/home/u027/project/software/bcbio/anaconda/bin:"$PATH" && /home/u027/project/software/bcbio/anaconda/bin/vep --vcf -o stdout -i /scratch/u027/project/analysis/wes_pilot/bcbio/work/test/gatk-haplotype/690.vcf.gz --fork 16 --species homo_sapiens --no_stats --cache --offline --dir /home/u027/project/software/bcbio/genomes/Hsapiens/hg38/vep --symbol --numbers --biotype --total_length --canonical --gene_phenotype --ccds --uniprot --domains --regulatory --protein --tsl --appris --af --max_af --af_1kg --af_esp --af_gnomad --pubmed --variant_class --allele_number --fasta /home/u027/project/software/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa.gz --plugin LoF,human_ancestor_fa:false,loftee_path:/home/u027/project/software/bcbio/anaconda/share/ensembl-vep-96.0-0 --plugin G2P,file:/home/u027/project/software/bcbio/genomes/Hsapiens/hg38/variation/DDG2P.30082018.csv --plugin MaxEntScan,/home/u027/project/software/bcbio/anaconda/share/maxentscan-0_2004.04.21-1 --plugin SpliceRegion --sift b --polyphen b --hgvsg --hgvs --shift_hgvs 1 --merged | sed '/^#/! s/;;/;/g' | bgzip -c > /scratch/u027/project/analysis/wes_pilot/bcbio/work/test/bcbiotx/tmpjfh3o9qe/690-vepeffects.vcf.gz
```

Changing the filename (and in hg38-resources.yaml) to G2P.csv doesn't fix the problem:

```
[ameynert@ultra test]$ head stdout_warnings.txt 
WARNING: Failed to instantiate plugin G2P: ERROR: No file specified or could not read from file file:/home/u027/project/software/bcbio/genomes/Hsapiens/hg38/variation/G2P.csv
```

**Comment by matthdsm** [comment]
well, beats me then..
@chapmanb, any more ideas?

M

**Comment by ameynert** [comment]
Is there a specific datatarget that should be installed? I wasn't getting a G2P.csv file when I ran this:

```
python bcbio_nextgen_install.py /home/u027/project/software/bcbio   --tooldir /home/u027/project/software/bcbio/tools   --genomes hg38 --aligners bwa   --datatarget variation --datatarget gnomad --datatarget cadd   --datatarget vep --datatarget dbnsfp --cores 16
```

**Comment by matthdsm** [comment]
Hi Allison,

No missing datatarget, just an undetected error in cloudbiolinux. The PR referenced above should fix that.
As for the error with the G2P plugin, still no idea..

M

**Comment by ameynert** [comment]
I just checked my installation logs, and the GGD recipe for genotype2phenotype didn't run:

```
[ameynert@ultra bcbio]$ grep 'GGD recipe' *.log 
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 seq 1000g-20150219_1
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 bwa 1000g-20150219
bcbio_install_20190529_0917.log:GGD recipe not available for hg38 rtg
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 ccds r20
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 capture_regions 20161202
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 coverage 2018-10-16
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 prioritize 20181227
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 dbsnp 151-20180418
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 hapmap_snps 20160105
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 1000g_omni_snps 20160105
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 ACMG56_genes 20160726
bcbio_install_20190529_0917.log:Running GGD recipe: hg38 1000g_snps 20160105
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 1000g_snps 20160105
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 mills_indels 20160105
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 1000g_indels 2.8_hg38_20150522
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 clinvar 20190513
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 qsignature 20160526
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 genesplicer 2004.04.03
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 effects_transcript 2017-03-16
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 varpon 20181105
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 vcfanno 20190119
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 viral 2017.02.04
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 gnomad 2.1
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 dbnsfp 3.5a
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA12878 v3_3_2
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA24385 v3_3_2
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA24631 v3_3_2
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 platinum-genome-NA12878 v2_0_1
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA12878-remap v2_19_1
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA12878-crossmap v2_19_1
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 dream-syn4-crossmap 2014-08-09
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 dream-syn3-crossmap 2014-08-04
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA12878-NA24385-somatic v3_3_2-2018-07-05
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA24143 v3_3_2
bcbio_install_20190529_1156.log:Running GGD recipe: hg38 giab-NA24149 v3_3_2
```

**Comment by ameynert** [comment]
I think I've fixed it. First, G2P.pm doesn't like the 'file:' prefix on my system. I've chopped that. Second, the switch for checking the file type, valid G2P format files don't necessarily have quotes around the tested 'allelic requirement' field in the header. I removed the double quotes from the regex.

Code snippet:

```
sub read_gene_data_from_file {
  my $self = shift;
  my $file = shift;
  my $delimiter = shift;
  my (@headers, %gene_data);

  if ($file =~ /^file\:(.+)/) {
    $file = $1;
  }

  my $assembly =  $self->{config}->{assembly};
  die("ERROR: No file specified or could not read from file ".($file || '')."\n") unless $file && -e $file;

  my @confidence_levels = @{$self->{user_params}->{confidence_levels}};

  # determine file type
  my $file_type;
  my $fh = FileHandle->new($file, 'r');
  while (<$fh>) {
    chomp;
      if (/Model_Of_Inheritance/) {
        $file_type = 'panelapp';
      } elsif (/allelic requirement/) {
        $file_type = 'g2p';
      } else {
	$file_type = 'unknown';
      }
      last;
  }
```

**Comment by matthdsm** [comment]
Hi Allison,
Nice work, could you submit this fix to the vep plugins repo? They'll be happy to have the fix.

Cheers
M

**Comment by ameynert** [comment]
I'll check to be sure I'm getting G2P results in the output VCF, then submit. Thanks for your patience!

**Comment by matthdsm** [comment]
Ensembl 97 was just released with an update for the G2P plugin. It should now function completely offline.
https://github.com/Ensembl/ensembl-vep/releases/tag/release%2F97.0

I'll open a PR with the required modifications once the update is accepted into conda.

Cheers
M

**Comment by matthdsm** [comment]
ref : https://github.com/bioconda/bioconda-recipes/pull/16185

**Comment by roryk** [comment]
Looks like this got merged in, thanks everyone! It is awesome having more folks contributing to figuring out and fixing problems!
